src/NuGet.Services.AzureSearch/Db2AzureSearch/Db2AzureSearchCommand.cs

```diff
@@ -0,0 +1,228 @@
+                    .ToList();
+                var allTasks = new[] { producerTask }.Concat(consumerTasks).ToList();
+
+                // If one of the tasks throws an exception before the work is completed, cancel the work.
+                var firstTask = await Task.WhenAny(allTasks);
+                if (firstTask.IsFaulted)
+                {
+                    cancelledCts.Cancel();
+                }
+
+                await firstTask;
+                await Task.WhenAll(allTasks);
+            }
+        }
+
+        private async Task InitializeAsync()
+        {
+            if (_options.Value.ReplaceIndexes)
+            {
+                await DeleteIndexIfExistsAsync(_options.Value.SearchIndexName);
+                await DeleteIndexIfExistsAsync(_options.Value.HijackIndexName);
+            }
+
+            await CreateIndexAsync<SearchDocument.Full>(_options.Value.SearchIndexName);
+            await CreateIndexAsync<HijackDocument.Full>(_options.Value.HijackIndexName);
+        }
+
+        private async Task DeleteIndexIfExistsAsync(string indexName)
+        {
+            if (await _serviceClient.Indexes.ExistsAsync(indexName))
+            {
+                _logger.LogWarning("Deleting index {IndexName}.", indexName);
+                await _serviceClient.Indexes.DeleteAsync(indexName);
+            }
+        }
+
+        private async Task CreateIndexAsync<T>(string indexName)
+        {
+            _logger.LogInformation("Creating index {IndexName}.", indexName);
+            await _serviceClient.Indexes.CreateAsync(new Index
+            {
+                Name = indexName,
+                Fields = FieldBuilder.BuildForType<T>(),
+            });
+        }
+
+        private async Task ProduceWorkAsync(
+            ConcurrentBag<NewPackageRegistration> allWork,
+            CancellationTokenSource completedCts,
+            CancellationToken cancellationToken)
+        {
+            await Task.Yield();
+            await _producer.ProduceWorkAsync(allWork, cancellationToken);
+            completedCts.Cancel();
+        }
+
+        private async Task ConsumeWorkAsync(
+            ConcurrentBag<NewPackageRegistration> allWork,
+            CancellationToken completedToken,
+            CancellationToken cancellationToken)
+        {
+            await Task.Yield();
+
+            var searchActions = new Queue<IndexAction<KeyedDocument>>();
+            var hijackActions = new Queue<IndexAction<KeyedDocument>>();
+
+            NewPackageRegistration work = null;
+            try
+            {
+                while ((allWork.TryTake(out work) || !completedToken.IsCancellationRequested)
+                    && !cancellationToken.IsCancellationRequested)
+                {
+                    // If there's no work to do, wait a bit before checking again.
+                    if (work == null)
+                    {
+                        await Task.Delay(TimeSpan.FromMilliseconds(100));
```

Comment:
is this a random wait time of 100ms or based of any analysis? just curious.

Reply:
Pretty much random. The idea is to quickly pick up new work as it is produced but not to try _constantly_.